code/iaas/auth-logic/src/main/java/io/cattle/platform/iaas/api/auth/dao/impl/AuthDaoImpl.java

```diff
@@ -129,70 +164,247 @@ public void updateAccount(Account account, String name, String kind, String exte
-        
-        if(null == userId || null == teamIds || null == orgIds) {
-            return new ArrayList<>();
-        }
-        
-        List<AccountRecord> projects = new ArrayList<>();
-        
-        Result<AccountRecord> orgProjects = create()
-                                    .selectFrom(ACCOUNT)
-                                    .where(ACCOUNT.EXTERNAL_ID.in(orgIds)
-                                            .and(ACCOUNT.EXTERNAL_ID_TYPE.eq("project:github_org")))
-                                            .and(ACCOUNT.STATE.ne("purged"))
-                                    .orderBy(ACCOUNT.ID.asc()).fetch();
-        
-        Result<AccountRecord> teamProjects = create()
-                                    .selectFrom(ACCOUNT)
-                                    .where(ACCOUNT.EXTERNAL_ID.in(teamIds)
-                                            .and(ACCOUNT.EXTERNAL_ID_TYPE.eq("project:github_team")))
-                                            .and(ACCOUNT.STATE.ne("purged"))
-                                    .orderBy(ACCOUNT.ID.asc()).fetch();
-        
-        for(AccountRecord orgProject: orgProjects) {
-            projects.add(orgProject);
-        }
-        
-        for(AccountRecord teamProject: teamProjects) {
-            projects.add(teamProject);
-        }
-        
-        projects.addAll(create()
-                .selectFrom(ACCOUNT)
-                .where(ACCOUNT.EXTERNAL_ID.eq(userId)
-                        .and(ACCOUNT.EXTERNAL_ID_TYPE.eq("project:github_user")))
-                        .and(ACCOUNT.STATE.ne("purged"))
-                        .fetch());
-        
+    public List<Account> getAccessibleProjects (Set<ExternalId> externalIds, boolean isAdmin, Long usingAccount) {
+        List<Account> projects = new ArrayList<>();
+        if(externalIds == null) {
+            return projects;
+        }
+        if (isAdmin) {
+            projects.addAll(create()
+                    .selectFrom(ACCOUNT)
+                    .where(ACCOUNT.KIND.eq(ProjectConstants.TYPE)
+                            .and(ACCOUNT.STATE.eq(CommonStatesConstants.ACTIVE)))
+                    .orderBy(ACCOUNT.ID.asc())
+                    .fetch());
+            return projects;
+        }
+        if (usingAccount != null) {
+            Account account = getAccountById(usingAccount);
+            if (account != null) {
+                if (account.getKind().equalsIgnoreCase(ProjectConstants.TYPE)) {
+                    projects.add(account);
+                }
+            }
+        }
+        //DSL.falseCondition is created so that we can dynamically build a or
+        //Condition without caring what the external Ids are and still make one
+        //Database call.
+        Condition allMembers = DSL.falseCondition();
+        for (ExternalId id: externalIds){
+            allMembers = allMembers.or(PROJECT_MEMBER.EXTERNAL_ID.eq(id.getId())
```

Comment:
I'm ok with this for the time being, but I'm pretty sure there's a limit to the # of condition clauses you can have in a SELECT statement.  (I do know there's already a limit to the # of values in an IN clause).
Once again, we can fix this later.  But just something to keep in mind.